ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/_layout/runs/$runIndex/rule-violations/index.tsx

```diff
@@ -18,34 +18,269 @@
+      </DialogHeader>
+      <div className='grid grid-cols-8 gap-2 text-sm'>
+        <div className='col-span-2 font-semibold'>License:</div>
+        <div className='col-span-6'>{details.ruleViolation.license}</div>
+        <div className='col-span-2 font-semibold'>License source:</div>
+        <div className='col-span-6'>{details.ruleViolation.licenseSource}</div>
+        <div className='col-span-2 font-semibold'>How to fix:</div>
+      </div>
+      <div className='flex whitespace-pre-line break-all text-sm'>
+        {details.ruleViolation.howToFix}
+      </div>
+      <DialogFooter>
+        <DialogClose asChild>
+          <Button type='button' variant='secondary'>
+            Close
+          </Button>
+        </DialogClose>
+      </DialogFooter>
+    </DialogContent>
+  );
+};
 
 const RuleViolationsComponent = () => {
+  const params = Route.useParams();
+  const search = Route.useSearch();
+
+  // Memoize the search parameters to prevent unnecessary re-rendering
+
+  const pageIndex = useMemo(
+    () => (search.page ? search.page - 1 : 0),
+    [search.page]
+  );
+
+  const pageSize = useMemo(
+    () => (search.pageSize ? search.pageSize : defaultPageSize),
+    [search.pageSize]
+  );
+
+  const groups = useMemo(
+    () => (search.groups ? search.groups : ['Severity']),
+    [search.groups]
+  );
+
+  const { data: ortRun } = useRepositoriesServiceGetOrtRunByIndexSuspense({
+    repositoryId: Number.parseInt(params.repoId),
+    ortRunIndex: Number.parseInt(params.runIndex),
+  });
+
+  const {
+    data: ruleViolations,
+    isPending,
+    isError,
+    error,
+  } = useRuleViolationsServiceGetRuleViolationsByRunId({
+    runId: ortRun.id,
+    // Fetch all data at once, as we need to do both grouping and
+    // pagination in front-end for consistency in data handling.
+    limit: 100000,
+    sort: 'rule',
+  });
+
+  const table = useReactTable({
+    data: ruleViolations?.data || [],
+    columns,
+    state: {
+      pagination: {
+        pageIndex,
+        pageSize,
+      },
+      grouping: groups,
+    },
+    getCoreRowModel: getCoreRowModel(),
+    getExpandedRowModel: getExpandedRowModel(),
+    getGroupedRowModel: getGroupedRowModel(),
+    getPaginationRowModel: getPaginationRowModel(),
+  });
+
+  if (isPending) {
+    return <LoadingIndicator />;
+  }
+
+  if (isError) {
+    toast.error('Unable to load data', {
+      description: <ToastError error={error} />,
+      duration: Infinity,
+      cancel: {
+        label: 'Dismiss',
+        onClick: () => {},
+      },
+    });
+    return;
+  }
+
   return (
-    <Card className='h-fit'>
-      <CardHeader>
-        <CardTitle>Rule violations</CardTitle>
-        <CardDescription>Rule violations of the project.</CardDescription>
-      </CardHeader>
-      <CardContent>
-        <Label className='font-semibold'>Not implemented yet.</Label>
-      </CardContent>
-    </Card>
+    <TooltipProvider>
```

Comment:
What is this TooltipProvider (or alternatively the one in DataTable) for?

Reply:
I had it in wrong place, I think it should be in the column component to provide a tooltip for the "more info" button.  In DataTable, I wanted to have tooltips for all grouped tables, to provide the user with info that a column can be toggled to group data.

This already stresses that we probably will have to wrap the whole app with TooltipProvider, to simplify.